Framework/CurveFitting/src/Functions/AsymmetricPearsonVII.cpp

```diff
@@ -0,0 +1,214 @@
+  const double peak_height = getParameter("PeakHeight");
+  const double peak_centre = getParameter("PeakCentre");
+  const double weight = 1.0 / getParameter("Width");
+  const double ml = getParameter("LeftShape");
+  const double mr = getParameter("RightShape");
+  double weight_sq = weight * weight;
+
+  // derivatives
+  for (size_t i = 0; i < nData; ++i) {
+    double offset = xValues[i] - peak_centre;
+    double offset_sq = offset * offset;
+    double left_denom, right_denom;
+
+    if (ml == 0.0)
+      left_denom = 2.0;
+    else
+      left_denom = 1.0 + 4.0 * offset_sq * (pow(2, 1.0 / ml) - 1.0) * weight_sq;
+
+    if (mr == 0.0)
+      right_denom = 2.0;
+    else
+      right_denom = 1.0 + 4.0 * offset_sq * (pow(2, 1.0 / mr) - 1.0) * weight_sq;
+
+    // derivative with respect to the peak_height
+    double left_ph_der, right_ph_der;
+
+    if (ml == 0.0)
+      left_ph_der = 1.0 / left_denom;
+    else
+      left_ph_der = 1.0 / pow(left_denom, ml);
+
+    if (mr == 0.0)
+      right_ph_der = 1.0 / right_denom;
+    else
+      right_ph_der = 1.0 / pow(right_denom, mr);
+
+    if (xValues[i] <= peak_centre)
+      out->set(i, 0, left_ph_der);
+    else
+      out->set(i, 0, right_ph_der);
+
+    // derivative with respect to the peak_centre
+    double left_pc_der, right_pc_der;
+
+    if (ml == 0.0)
+      left_pc_der = 0.0;
+    else
+      left_pc_der = 8.0 * peak_height * ml * offset * (pow(2, 1.0 / ml) - 1.0) * weight_sq / pow(left_denom, 1.0 + ml);
+
+    if (mr == 0.0)
+      right_pc_der = 0.0;
+    else
+      right_pc_der =
+          8.0 * peak_height * mr * offset * (pow(2, 1.0 / mr) - 1.0) * weight_sq / pow(right_denom, 1.0 + mr);
+
+    if (xValues[i] <= peak_centre)
+      out->set(i, 1, left_pc_der);
+    else
+      out->set(i, 1, right_pc_der);
+
+    // derivative with respect to the weight, not width
+    double left_weight_der, right_weight_der;
+    if (ml == 0.0)
+      left_weight_der = 0.0;
+    else
+      left_weight_der =
+          -8.0 * peak_height * ml * offset_sq * (pow(2, 1.0 / ml) - 1.0) * weight / pow(left_denom, 1.0 + ml);
+
+    if (mr == 0.0)
+      right_weight_der = 0.0;
+    else
+      right_weight_der =
+          -8.0 * peak_height * mr * offset_sq * (pow(2, 1.0 / mr) - 1.0) * weight / pow(right_denom, 1.0 + mr);
+
+    if (xValues[i] <= peak_centre)
+      out->set(i, 2, left_weight_der);
+    else
+      out->set(i, 2, right_weight_der);
+
+    // derivative with respect to ml
+    double ml_der;
+    if (ml == 0.0)
+      ml_der = -peak_height / 2.0 * log(4.0 * offset_sq * weight_sq);
+    else
+      ml_der = peak_height *
+               (4.0 * offset_sq * weight_sq * pow(2, 1.0 / ml) * log(2.0) / ml / left_denom - log(left_denom)) /
+               pow(left_denom, ml);
+
+    if (xValues[i] <= peak_centre)
+      out->set(i, 3, ml_der);
+    else
+      out->set(i, 3, 0.0);
+
+    // derivative with respect to mr
```

Comment:
doing each derivative individually like this is clearer. However, it results in a long list of if statements, which will impact the performance of the code. You could do it to be something like:
```
if (ml==0.0){
ddml = getDerivativeWRTmlEqual0(....);
ddw = 0.0;
....}
else{
ddml = getDerivativeWRTml(....);
ddw = getDerivativeWRTw(....);
}
```

Reply:
right, I've fixed that